url9a01f43c15_poragchowdhury_securitygame_tgz-pJ8-securitygame_GameMasterJ8/src/securitygame/AttackerMonitor.java

```diff
@@ -0,0 +1,353 @@
+            String publicString = "6," + publicNodes.get(i).getNodeID()+",";
+            ArrayList<Node> neighbors = publicNodes.get(i).getNeighborList();
+            for(int j = 0; j < neighbors.size(); j++)
+                publicString += neighbors.get(j).getNodeID() + ",";
+                publicString = publicString.substring(0, publicString.length() - 1);
+                history.println(publicString);
+            }
+            history.close();
+        } catch (IOException e) {e.printStackTrace();}
+        history.close();
+    }
+
+    /**
+     * Parses attacker's move from .attack file
+     */
+    public void readMove(){
+		try{
+			visibleNet = Parser.parseAttackerHistory(attackerName, defenderName, graphName);
+			history = new PrintWriter(new FileWriter(attackerName + "-" + defenderName + "-" + graphName + ".history", true));
+			File csv = new File(attackerName + "-" + defenderName + "-" + graphName + ".attack");
+			CSVParser parser = CSVParser.parse(csv, StandardCharsets.US_ASCII, CSVFormat.DEFAULT);
+			
+			for(CSVRecord csvRecord : parser){
+				Iterator<String> itr = csvRecord.iterator();
+                int mode = Integer.parseInt(itr.next());
+                int id;
+                String neighbors = ",";
+                switch (mode){
+                	case 0://attack node
+                    id = Integer.parseInt(itr.next());
+                    if(isAvailableNode(id) && isValidAttack(id))
+                    {
+                        budget -= Parameters.ATTACK_RATE;
+                        Node n = net.getNode(id);
+                        int attackRoll = r.nextInt(Parameters.ATTACK_ROLL) + 1; 
+                        if(attackRoll >= n.getSv()){
+                        	visibleNet.getNode(id).setPv(n.getPv());
+                        	visibleNet.getNode(id).setSv(n.getSv());
+                        	visibleNet.getNode(id).setHoneyPot(n.getHoneyPot());
+                        	visibleNet.getNode(id).setCaptured(true);
+                        	for(int j = 0; j < n.neighbor.size(); j++){
+                        		visibleNet.getNode(id).addNeighbor(visibleNet.getNode(n.neighbor.get(j).getNodeID()));
+                        		neighbors += n.neighbor.get(j).getNodeID() + ",";
+                        	}
+                        	neighbors = neighbors.substring(0, neighbors.length()-1);
+                        	System.out.println("attack on node " + id + " was successful with a roll of " + attackRoll + "!");
+                            points += visibleNet.getNode(id).getPv();
+                        	history.println("0," + id + ",true," + attackRoll + "," + n.getPv() + "," + n.getSv() + "," + n.getHoneyPot() + neighbors);
+                        }else{
+                        	System.out.println("attack on node " + id + " was unsuccessful with a roll of " + attackRoll);
+                        	history.println("0," + id + ",false," + attackRoll);
+                        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 101:117, 133:149.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.